

library(shiny)
library(shinydashboard)

### Define UI ----
ui <- dashboardPage(
    dashboardHeader(title = "UHR-IonStar v1.4"),
    dashboardSidebar(
        sidebarMenu(
            menuItem("Homepage", tabName = "Homepage", icon = icon("home")),
            menuItem("IonStarStat", tabName = "IonStarStat", icon = icon("edit")),
            menuItem("Processing", tabName = "Process", icon = icon("edit")),
            menuItem("Visualization", tabName = "Visual", icon = icon("eye")),
            menuItem("Changes Discovery", tabName = "Biomarkers", icon = icon("filter")),
            menuItem('Re-verification', tabName = "Verify", icon = icon("check"))

        )
    ),
    dashboardBody(
        tags$head(
            tags$style(
                "body{
          min-height: 800px;
          height: auto;
          min-width: 1600px;
          margin: auto;
}"
            )
        ),

        tabItems(
            ##Homepage
            {
                tabItem(tabName = "Homepage",
                        fluidPage(
                            titlePanel(strong("UHR-IonStar")),
                            mainPanel(
                                img(src="logo.png",height=120,width=120),
                                h4(strong("UHR-IonStar is a Shiny-based interactive web app designed for processing, visualization and mining of quantitative proteomics data generated by IonStar")),
                                h4("Fundamental components of UHR-IonStar include:"),
                                h4(icon("check"),"Frame generation and protein quantification"),
                                h4(icon("check"),"Data cleanup"),
                                h4(icon("check"),"Case-control ratio calculation and statistical testing"),
                                h4(icon("check"),"Graphic depiction of quantitative results"),
                                h4(icon("check"),"Basic data mining (e.g. Principle Component Analysis)"),
                                h4(icon("check"),"DAVID-based Gene Ontology analysis"),
                                h4("To get started, click on the blocks in left-side tab then upload files you want to process."),
                                h4("Click the botton below to download the manual:"),
                                downloadButton("downloadManual", label="Download Manual", icon=icon("download")),
                                h4("For questions, suggestions, and other relevant topics about UHR-IonStar, please contact the developers:"),
                                h4("Shuo Qian:",
                                   "sqian@buffalo.edu"),
                                h4("Shichen Shen:",
                                   "shichens@buffalo.edu"),
                                h4("Jun Qu:",
                                   "junqu@buffalo.edu")
                            )
                        )
                )},

            #IonStarStat
            {
                tabItem(tabName = "IonStarStat",
                        fluidPage(
                            titlePanel(strong("IonStarStat")),
                            sidebarLayout(
                                position = "left",
                                sidebarPanel(
                                    h4(strong("Frames Generation:")),
                                    fileInput("StatUploadSEIVE", "Upload SIEVE database (.sdb)",
                                              accept=c(".sdb")),
                                    helpText('Please upload spectrum report after you see "Upload complete" on the bar above!', style = 'color:blue'),
                                    fileInput("StatUploadSR", "Upload Spectrum report (.csv)",
                                              accept=c("text/csv",
                                                       "text/comma-separated-values,text/plain",
                                                       ".csv")),
                                    actionButton(inputId = "StatGoButton1",label = "Submit database and spectrum report",icon=icon("pencil"),width="100%"),
                                    span(textOutput("StatNotice1"), style = 'color:green'),
                                    br(),
                                    h6(strong("Several settings in Spectrum report:")),
                                    textInput(inputId = "StatFileNumber",
                                              label = "Input the number of column which contains filename information:", value = "1"),
                                    textInput(inputId = "StatScanNumber",
                                              label = "Input the number of column which contains MS2 scan number:", value = "3"),
                                    textInput(inputId = "StatProAccession",
                                              label = "Input the number of column which contains protein accessions:", value = "16"),
                                    textInput(inputId = "StatPepSeq",
                                              label = "Input the number of column which contains peptide sequence:", value = "15"),
                                    actionButton(inputId = "StatGoButton2",label = "Generate Frames",icon=icon("pencil"),width="100%"),
                                    span(textOutput("StatFrameNotice"), style = 'color:green'),
                                    br(),
                                    downloadButton("StatFrameDownload", label="Download Annotated Frame List", icon=icon("download")),
                                    downloadButton("StatSampleIDDownload", label="Download Sample ID", icon=icon("download")),
                                    br(),
                                    tags$hr(),
                                    h4(strong("Protein Quantification:")),
                                    fileInput("StatUploadFrame", "Upload Frame List(.csv)",
                                              accept=c("text/csv",
                                                       "text/comma-separated-values,text/plain",
                                                       ".csv")),
                                    fileInput("StatUploadSampleID", "Upload Sample ID (.csv)",
                                              accept=c("text/csv",
                                                       "text/comma-separated-values,text/plain",
                                                       ".csv")),
                                    actionButton(inputId = "StatGoButton3",label = "Submit Annotated Framelist and SampleID",icon=icon("pencil"),width="100%"),
                                    br(),
                                    span(textOutput("StatNotice2"), style = 'color:green'),
                                    br(),
                                    radioButtons("Norm_method","Select inter-group peptide normalization method:",
                                                 c("Total Ion Currents"="TIC",
                                                   "Quantiles"="quantiles",
                                                   "No normalization" = 'none')),
                                    radioButtons("Quan_method","Select method that aggregates peptides to proteins",
                                                 c("The Sum of Intensities"="sum",
                                                   "General Linear Mixed Model (GLMM)"="fit")),
                                    checkboxGroupInput("Outlier_rejection",
                                                       "Choose whether you want to do outlier rejection by PCout method:",
                                                       choices = "Outlier Rejection"),
                                    checkboxGroupInput("Share_rm",
                                                       "Choose whether you want to remove shared frames between two or more different species:",
                                                       choices = "Mixed Proteomes Deconvolution"),
                                    actionButton(inputId = "StatGoButton4",label = "Perform Protein Quantification",icon=icon("pencil"),width="100%"),
                                    span(textOutput("noticeStat8"),style = 'color:green'),
                                    br(),
                                    downloadButton("downloadDataStat4", label="Download Quantification Data", icon=icon("download")),
                                    downloadButton("downloadDataStat5", label="Download Peptides", icon=icon("download")),
                                    br(),
                                    tags$hr()

                                ),#sidebarpanel
                                mainPanel(
                                    h3(strong("IonStarStat")),
                                    textOutput("message1"),
                                    textOutput("message2")
                                )
                            )#sidebarlayout
                        )#fieldpage
                )}, # tabItem

            #Data processing
            {tabItem(tabName = "Process",
                     fluidPage(
                         titlePanel(strong("Data Processing")),
                         sidebarLayout(
                             position = "left",
                             sidebarPanel(
                                 fileInput("file1", "Upload quantitative results (.csv):",
                                           accept=c("text/csv",
                                                    "text/comma-separated-values,text/plain",
                                                    ".csv")),
                                 fileInput("file2", "Upload group file (.csv):",
                                           accept=c("text/csv",
                                                    "text/comma-separated-values,text/plain",
                                                    ".csv")),
                                 tags$hr(),
                                 radioButtons("qpat","The pattern of the quantitative file:",
                                              c("Protein file"="pro",
                                                "Peptide file"="pep")),
                                 actionButton(inputId = "goButton",label = "Submit",icon=icon("upload"),width="100%"),


                                 br(),
                                 tags$hr(),

                                 helpText("Please fill in your control group shown in the right, case sensitive (default is the first one):", style = 'color:blue'),
                                 #  selectInput(inputId = "ctrlGroup", label = "Assign control group", choices = " "),
                                 textInput(inputId = "ctrlGroup",
                                           label = "Assign control group", value="XXX"),
                                 radioButtons("testing","Statistical testing method",
                                              c("Original t-test"="t",
                                                "Paired t-test"="paired-t")),
                                 textInput("decoy", "Decoy Protein identifier",
                                           value="XXX"),
                                 actionButton("calculate",label="Start Data Processing",icon=icon("pencil"),width="100%"),
                                 span(textOutput("notice1"),style = 'color:green'),
                                 downloadButton("downloadData", label="Download Processed Quantity Results", icon=icon("download")),
                                 downloadButton("downloadData2", label="Download Average Ratios", icon=icon("download"),width="100%"),
                                 downloadButton("downloadData_all", label="Download All Together", icon=icon("download"),width="100%"),
                                 helpText("Note: Decoy entries and proteins with missing data will be removed automatically during the data processing step.", style = 'color:blue')
                             ),
                             mainPanel(
                                 h4(strong("Quantitative Results")),
                                 DT::dataTableOutput("contents"),
                                 textOutput("submitted")
                             )
                         )
                     )
            )}, # tabItem

            ##Data visualization
            {tabItem(tabName = "Visual",
                     fluidPage(
                         titlePanel(strong("Data Visualization")),
                         sidebarLayout(
                             position = "left",
                             sidebarPanel(
                                 selectInput(inputId = "plotType",
                                             label = "Select plot type:",
                                             choices = c("Intra-group CV boxplot", "Protein intensity-rank curve", "Inter-group correlation plot",
                                                         "Pearson correlation matrix plot","Principal Component Analysis (PCA)")),
                                 p(strong("*For Pearson correlation plot ONLY:")),
                                 helpText("Please fill the lower bound of Correlation value showing in the plot(number only):", style = 'color:blue'),
                                 textInput(inputId = "corr",
                                           label = "Correlation Lower Bound", value = "0.9"),
                                 p(strong("*For Inter-group correlation plot ONLY:")),
                                 helpText("Please fill in the groups you want to see, case sensitive ( Default is the first two groups):", style = 'color:blue'),
                                 textInput(inputId = "inter_group1",
                                           label = "Group 1", value = "XXX"),
                                 textInput(inputId = "inter_group2",
                                           label = "Group 2", value = "XXX"),

                                 actionButton("plot",label="Plot",icon=icon("paint-brush"),width="100%"),
                                 p(strong("Plot saving options:")),
                                 helpText("Recommends larger width and height values when plots for Pearson correlation matrix and PCA.", style = 'color:blue'),
                                 sliderInput("sliderwid","Plot width",
                                             min = 4,max = 20,value = 6),
                                 sliderInput("sliderhei","Plot height",
                                             min = 4,max = 20,value = 4),
                                 #   checkboxGroupInput("exporttype","Export format",
                                 #                       choices = list("Vector-PDF" = 1,
                                 #                                      "Bitmap-PNG" = 2,
                                 #                                      "Bitmap-TIFF" = 3)),
                                 downloadButton("downloadPlot", label="Download Current Plot",
                                                icon=icon("download"),width="100%"),
                                 br(),
                                 h4(strong("Ratio distribution plots settings:")),
                                 checkboxGroupInput("ratio_plot_selection",
                                                    "Choose group/groups you want to show in ratio distribution plot:",
                                                    choices = " ", selected = " "),
                                 checkboxGroupInput("dist_correction",
                                                    "Choose whether you want to do distribution correction:",
                                                    choices = "Correction"),
                                 actionButton("plot2",label="Plot for Ratio Distribution",
                                              icon=icon("paint-brush"),width="100%"),
                                 sliderInput("sliderwid2","Plot width",
                                             min = 4,max = 20,value = 6),
                                 sliderInput("sliderhei2","Plot height",
                                             min = 4,max = 20,value = 4),
                                 downloadButton("downloadPlot2", label="Download Current Plot",
                                                icon=icon("download"),width="100%"),
                                 downloadButton("downloadData4", label = "Download corrected ratio dataset",
                                                icon = icon("download"), width = "100%")

                             ),
                             mainPanel(
                                 h4(strong("Plot Area")),
                                 plotOutput("Plotting"),
                                 DT::dataTableOutput("setup"),
                                 h4(strong("Area for Ratio Distribution Plots")),
                                 plotOutput("Plotting2")
                             )
                         )
                     )
            )},# tabItem

            ## Discovery of Changes
            {tabItem(tabName = "Biomarkers",
                     fluidPage(
                         titlePanel(strong("Discovery of Changes")),
                         sidebarLayout(
                             position = "left",
                             sidebarPanel(
                                 p(strong("Settings for finding significantly changed proteins")),
                                 selectInput(inputId = "select_group",
                                             label = "Select group",
                                             choices = ""),
                                 textInput("R_cutoff", "Set ratio cutoff",
                                           value=1.4),
                                 textInput("p_cutoff", "Set p-value cutoff",
                                           value=0.05),
                                 actionButton(inputId = "upload_cutoff",label = "Submit cutoff settings",
                                              icon=icon("upload"),width="100%"),
                                 downloadButton("downloadData3", label="Download Significantly Changed Data",
                                                icon=icon("download")),
                                 selectInput(inputId = "plotType2",
                                             label = "Select plot type:",
                                             choices = c("Valcano Plot",
                                                         "Intensity curve with changes highlighted",
                                                         "Plot for Up-regulated","Plot for Down-regulated","Gene Ontology (Accession)",
                                                         "Gene Ontology (ProteinID)")),
                                 helpText("*For up-/down-regulated proteins plot only:", style = 'color:blue'),
                                 textInput("TOPx",label = "Assign how many proteins shown in the plot:",value = 50),
                                 actionButton("plot3",label="Plot",icon=icon("paint-brush"),width="100%"),
                                 sliderInput("sliderwid3","Plot width",
                                             min = 4,max = 20,value = 6),
                                 sliderInput("sliderhei3","Plot height",
                                             min = 4,max = 20,value = 4),
                                 downloadButton("downloadPlot3", label="Download Current Plot",
                                                icon=icon("download"))

                             ),
                             mainPanel(
                                 h4(strong("Table Area")),
                                 DT::dataTableOutput("sign_contents"),
                                 h4(strong("Plot Area")),
                                 plotOutput("Plotting3")
                             )
                         )
                     )
            ) #tabItem
            },

            #Re-Verification
            {tabItem(tabName = "Verify",
                     fluidPage(
                         titlePanel(strong("Re-verification")),
                         sidebarLayout(
                             position = "left",
                             sidebarPanel(
                                 fileInput("veri_file1", "Repload Protein Quantification result (.csv):",
                                           accept=c("text/csv",
                                                    "text/comma-separated-values,text/plain",
                                                    ".csv")),
                                 fileInput("veri_file2", "Repload Peptide Quantification result (.csv):",
                                           accept=c("text/csv",
                                                    "text/comma-separated-values,text/plain",
                                                    ".csv")),
                                 tags$hr(),
                                 actionButton(inputId = "veri_submit",label = "Submit Protein and Peptide Outputs",icon=icon("upload"),width="100%"),
                                 span(textOutput("veri_notice1"),style = 'color:green'),
                                 helpText("*Note: double-check if the groups(columns) in both protein and peptide files are with the same order.", style = 'color:blue'),
                                 actionButton(inputId = "veri_inge",label = "Start Integration",icon=icon("pencil"),width="100%"),
                                 span(textOutput("veri_notice2"),style = 'color:green'),
                                 downloadButton("veri_downloadData1", label="Download Integrated Quantity Results", icon=icon("download")),
                                 br(),
                                 tags$hr(),

                                 helpText("Please fill the protein accession you want to see (case sensitive):", style = 'color:blue'),
                                 #  selectInput(inputId = "ctrlGroup", label = "Assign control group", choices = " "),
                                 textInput(inputId = "veri_accession",
                                           label = "Protein Accession", value="ABCD01234"),
                                 helpText("The legend won't show in the plot area if the number of peptide is bigger than 5.", style = 'color:blue'),
                                 helpText("The first bar in each group is the protein ratio value.", style = 'color:blue'),
                                 helpText("The download one will contain the legends. Larger width and height values recommended.", style = 'color:blue'),
                                 actionButton("veri_visual",label="Show Target Protein Accession and Its Peptides",icon=icon("image"),width="100%"),
                                 br(),
                                 sliderInput("veri_sliderwid1","Plot width",
                                             min = 10,max = 30,value = 15),
                                 sliderInput("veri_sliderhei1","Plot height",
                                             min = 10,max = 30,value = 15),
                                 downloadButton("veri_downloadPlot1", label="Download Current Plot with legends", icon=icon("download"),width="100%")
                             ),
                             mainPanel(
                                 h4(strong("Plot Area")),
                                 plotOutput("Veri_plot")
                             )
                         )
                     )
            )} # tabItem


        ) # tabItems
    ) #dashboardBody
) #dashboardPage


### Define server logic ----
server <- function(input, output, session) {
    #### IonStarStat ####
    ## Frame Generation##

    output$StatNotice1 <- renderText({
        input$StatGoButton1
        Database_dir <<- isolate(input$StatUploadSEIVE)
        if (is.null(Database_dir))
            return(NULL)

        Sprepo_dir <<- isolate(input$StatUploadSR)
        if (is.null(Sprepo_dir))
            return(NULL)

        print('Submission completed.')
    })


    dbStat <- reactive({dbConnect(dbDriver("SQLite"),dbname = Database_dir$datapath)})
    spStat <- reactive({read.csv(Sprepo_dir$datapath)})

    testStat2 <- eventReactive(input$StatGoButton2,FrameGen(dbStat(),spStat(),input$StatFileNumber,input$StatScanNumber,
                                                            input$StatProAccession, input$StatPepSeq))
    output$StatFrameNotice <- renderText({
        testStat2()$notice
    })
    output$message1 <- renderText({
        testStat2()$message1
    })
    output$message2 <- renderText({
        testStat2()$message2
    })

    output$StatFrameDownload <- downloadHandler(
        filename = function() {
            paste('Annotated_Frames-', gsub(" ","-",date()), '.csv', sep='')
        },
        content = function(con) {
            write.csv(testStat2()$framelist, row.names = FALSE, con)
        }
    )
    output$StatSampleIDDownload <- downloadHandler(
        filename = function() {
            paste('Sample_ID-', gsub(" ","-",date()), '.csv')
        },
        content = function(con) {
            write.csv(testStat2()$sampleid, con, sep='', row.names = FALSE)
        }
    )

    ## Protein Quantification ##

    output$StatNotice2 <- renderText({
        input$StatGoButton3
        FrameList_dir <<- isolate(input$StatUploadFrame)
        if (is.null(FrameList_dir))
            return(NULL)

        SampleID_dir <<- isolate(input$StatUploadSampleID)
        if (is.null(SampleID_dir))
            return(NULL)

        Database_dir <<- isolate(input$StatUploadDatabase)
        if (is.null(Database_dir))
            return(NULL)

        print('Submission completed.')
    })

    framelist <- reactive({read.csv(FrameList_dir$datapath, header = TRUE, check.names = FALSE, stringsAsFactors = FALSE)})
    sampleid <- reactive({read.csv(SampleID_dir$datapath, header = TRUE, check.names = FALSE)})
    database <- reactive({read.fasta(Database_dir$datapath, seqtype = "AA", as.string = TRUE)})

    testStat3 <- eventReactive(input$StatGoButton4,ProQuant(framelist(),sampleid(),database(),rm = input$Share_rm,
                                                            norm_method = input$Norm_method, quan_method = input$Quan_method,
                                                            outlier = input$Outlier_rejection))
    output$noticeStat8 <- renderText({
        testStat3()$notice
    })


    output$downloadDataStat4 <- downloadHandler(
        filename = function() {
            paste('Quan_data-', gsub(" ","-",date()), '.csv', sep='')
        },
        content = function(con) {
            write.csv(testStat3()$quan, con)
        }
    )
    output$downloadDataStat5 <- downloadHandler(
        filename = function() {
            paste('Pep_data-', gsub(" ","-",date()), '.csv', sep='')
        },
        content = function(con) {
            write.csv(testStat3()$pep, con)
        }
    )
    output$downloadDataStat6 <- downloadHandler(
        filename = function() {
            paste('Frame-Shared-', gsub(" ","-",date()), '.csv', sep='')
        },
        content = function(con) {
            write.csv(testStat3()$OringalFrame, con)
        }
    )



    #### Processing ####
    {

        output$downloadManual <- downloadHandler(
            filename <- function(){
                paste("Manual","html",sep = ".")
            },

            content <- function(file){
                file.copy("Manual.html",file)
            },
            contentType = "application/html"
        )



        ## input quan_file and group_file
        output$contents <- DT::renderDataTable({
            input$goButton
            quan_file_dir <<- isolate(input$file1)
            if (is.null(quan_file_dir))
                return(NULL)

            group_file_dir <<- isolate(input$file2)
            if (is.null(group_file_dir))
                return(NULL)

            tmp<-read.csv(group_file_dir$datapath, header = TRUE,row.names = 1)
            tmp2 <- as.matrix(unique(as.character(tmp$GroupID)))
            colnames(tmp2)<- "Groups"
            tmp2
        })


        quan_file <- reactive({read.csv(quan_file_dir$datapath, header = TRUE)})
        group_file <- reactive({read.csv(group_file_dir$datapath, header = TRUE)})

        observeEvent(input$calculate,{
            decoy_id <<- as.character(input$decoy)

        })
        #IonStar_Ratio_p.R

        test <- eventReactive(input$calculate,
                              Ratio_p(quan_file(),group_file(),
                                      decoy = decoy_id, control_group = input$ctrlGroup,
                                      is_t_test = input$testing, pro_or_pep = input$qpat))

        output$notice1 <- renderText({
            test()$notice
        })

        output$downloadData <- downloadHandler(
            filename = function() {
                paste('quan_results-', gsub(" ","-",date()), '.csv', sep='')
            },
            content = function(con) {
                write.csv(test()$quan_results, con, row.names = FALSE)
            }
        )
        output$downloadData2 <- downloadHandler(
            filename = function() {
                paste('ave_ratios-', gsub(" ","-",date()), '.csv', sep='')
            },
            content = function(con) {
                write.csv(test()$ave_ratio, con, row.names = FALSE)
            }
        )
        output$downloadData_all <- downloadHandler(
            filename = function() {
                paste('quan&ratios-', gsub(" ","-",date()), '.csv', sep='')
            },
            content = function(con) {
                write.csv(test()$all, con, row.names = FALSE)
            }
        )

    }
    #### Visualization ####

    # plot for all
    test2 <- eventReactive(input$plot, plot_for_all(plot_selection=input$plotType,
                                                    Group1 = input$inter_group1, Group2= input$inter_group2, corr = input$corr))
    output$Plotting <- renderPlot({
        test2()
    })

    output$downloadPlot <- downloadHandler(
        filename = function() {
            paste('Plot-', gsub(" ","-",date()),'.png', sep='')
        },
        content = function(con4) {
            device <- function(..., width, height) grDevices::png(..., width = input$sliderwid, height = input$sliderhei, res = 300, units = "in")
            ggsave(con4, plot=test2(),device = device)
        }
    )

    #update seletions
    observe({
        input$calculate
        x <- test()$group_char[test()$group_char != test()$control_group]

        if (is.null(x))
            x <- character(0)

        updateCheckboxGroupInput(session, "ratio_plot_selection",
                                 label = "Choose group/groups you want to show in ratio distribution plot:",
                                 choices = x,
                                 selected = x[1]
        )
        updateSelectInput(session, inputId = "select_group",
                          label = "Select group",
                          choices = x)
    })

    # ratio distribution (density) plot
    test3 <- eventReactive(input$plot2, plot_ratio_distribution(
        selected = input$ratio_plot_selection, dist_corr = input$dist_correction))
    output$Plotting2 <- renderPlot({
        test3()$plotting
    })

    output$downloadPlot2 <- downloadHandler(
        filename = function() {
            paste('ratio_plot-', gsub(" ","-",date()), '.png', sep='')
        },
        content = function(con2) {
            device <- function(..., width, height) grDevices::png(..., width = input$sliderwid2, height = input$sliderhei2, res = 300, units = "in")
            ggsave(con2, plot=test3()$plotting,device = device)
        }
    )

    output$downloadData4 <- downloadHandler(
        filename = function() {
            paste('corrected_ratios-', gsub(" ","-",date()), '.csv', sep='')
        },
        content = function(con) {
            write.csv(test3()$correction_ratio, con)
        }
    )

    #########Discovery of Changes#############
    # finding significantly changed proteins
    test4 <- eventReactive(input$upload_cutoff, sign_changes(input$select_group,
                                                             input$R_cutoff, input$p_cutoff))
    output$sign_contents <- DT::renderDataTable({
        input$upload_cutoff
        test4()$quan_changed
    })
    output$downloadData3 <- downloadHandler(
        filename = function() {
            paste('sign-changed', gsub(" ","-",date()), '.csv', sep='')
        },
        content = function(con3) {
            write.csv(test4()$quan_changed, con3)
        }
    )
    # plot for all 2
    test5 <- eventReactive(input$plot3,plot_for_all2(plot_selection = input$plotType2,
                                                     TOPx = input$TOPx))
    output$Plotting3 <- renderPlot({
        test5()
    })
    output$downloadPlot3 <- downloadHandler(
        filename = function() {
            paste('sign_plot-', gsub(" ","-",date()), '.png', sep='')
        },
        content = function(con3) {
            device <- function(..., width, height) grDevices::png(..., width = input$sliderwid3, height = input$sliderhei3, res = 300, units = "in")
            ggsave(con3, plot=test5(),device = device)
        }
    )

    ###############Re-Verification######################
    output$veri_notice1 <- renderText({
        input$veri_submit
        Protein_dir <<- isolate(input$veri_file1)
        if (is.null(Protein_dir))
            return(NULL)

        Peptide_dir <<- isolate(input$veri_file2)
        if (is.null(Peptide_dir))
            return(NULL)

        print('Submission completed.')
    })

    veri_protein <- reactive({read.csv(Protein_dir$datapath, header = TRUE)})
    veri_peptide <- reactive({read.csv(Peptide_dir$datapath, header = TRUE)})

    #ProteinFilter.R

    veri_test1 <- eventReactive(input$veri_inge,ProPepCombine(veri_protein(),veri_peptide()))
    output$veri_notice2 <- renderText({
        veri_test1()$notice
    })

    output$veri_downloadData1 <- downloadHandler(
        filename = function() {
            paste('Inte_data-', gsub(" ","-",date()), '.csv', sep='')
        },
        content = function(con) {
            write.csv(veri_test1()$df, con)
        }
    )

    veri_test2 <- eventReactive(input$veri_visual,CustomSearchPlot(veri_test1()$df,
                                                                   input$veri_accession))
    output$Veri_plot <- renderPlot({
        veri_test2()$plot2
    })

    output$veri_downloadPlot1 <- downloadHandler(
        filename = function() {
            paste('Accession_plot-', gsub(" ","-",date()), '.png', sep='')
        },
        content = function(con5) {
            device <- function(..., width, height) grDevices::png(..., width = input$veri_sliderwid1, height = input$veri_sliderhei1, res = 300, units = "in")
            ggsave(con5, plot=veri_test2()$plot1,device = device)
        }
    )

}

# Run the application
shinyApp(ui = ui, server = server)
